src/restconfig/src/test/java/org/geoserver/rest/resources/ResourceTest.java

```diff
@@ -0,0 +1,310 @@
+                + "\"parent\":   {"
+                + "  \"path\": \"/\","
+                + "    \"link\":     {"
+                + "      \"href\": \"http://localhost:8080/geoserver/rest/resources/?format=json\","
+                + "      \"rel\": \"alternate\","
+                + "      \"type\": \"application/json\""
+                + "  }"
+                + "},"
+                + "\"lastModified\": \"" + FORMAT.format(myRes.lastmodified()) + "\","
+                + "  \"children\": {\"child\": [  {"
+                + "    \"name\": \"myres\","
+                + "    \"link\":     {"
+                + "      \"href\": \"http://localhost:8080/geoserver/rest/resources/mydir/myres\","
+                + "      \"rel\": \"alternate\","
+                + "      \"type\": \"application/octet-stream\""
+                + "    }"
+                + "  }]}"
+                + "}}";
+        JSONAssert.assertEquals(expected, (JSONObject) json);
+    }
+    
+    @Test
+    public void testDirectoryHTML() throws Exception {
+        XMLUnit.setXpathNamespaceContext(NS_HTML);
+        Document doc = getAsDOM("/rest/resource/mydir?format=html");
+        //print(doc);
+        XMLAssert.assertXpathEvaluatesTo("Name: 'mydir'", "/x:html/x:body/x:ul/x:li[1]", doc);
+        XMLAssert.assertXpathEvaluatesTo("http://localhost:8080/geoserver/rest/resources/?format=html", 
+                "/x:html/x:body/x:ul/x:li[2]/x:a/@href", doc);
+        XMLAssert.assertXpathEvaluatesTo("Last modified: " + new Date(myRes.lastmodified()).toString(), 
+                "/x:html/x:body/x:ul/x:li[3]", doc);
+        XMLAssert.assertXpathEvaluatesTo("http://localhost:8080/geoserver/rest/resources/mydir/myres", 
+                "/x:html/x:body/x:ul/x:li[4]/x:ul/x:li/x:a/@href", doc);
+    }
+    
+    @Test
+    public void testDirectoryHeaders() throws Exception {
+        MockHttpServletResponse response = getAsServletResponse("/rest/resource/mydir?format=xml");
+        Assert.assertEquals(FORMAT_HEADER.format(myRes.lastmodified()), response.getHeader("Last-modified"));
+        Assert.assertEquals("/", response.getHeader("Resource-parent"));
+        Assert.assertEquals("directory", response.getHeader("Resource-type"));
+        Assert.assertEquals("application/xml", response.getHeader("Content-type"));
+    }
+    
+    @Test
+    public void testDirectoryMimeTypes() throws Exception {
+        XMLUnit.setXpathNamespaceContext(NS_XML);
+        Document doc = getAsDOM("/rest/resource/mydir2?format=xml");
+        //print(doc);
+        XMLAssert.assertXpathEvaluatesTo("image/png", "/ResourceDirectory/children/child[name='imagewithoutextension']/atom:link/@type", doc);
+        XMLAssert.assertXpathEvaluatesTo("image/png", "/ResourceDirectory/children/child[name='fake.png']/atom:link/@type", doc);
+    }
+        
+    @Test
+    public void testUpload() throws Exception {
+        put("/rest/resource/mydir/mynewres", "This is my new test.");
```

Comment:
@NielsCharlier great opportunity to include non-ASCII characters to see if they survive the round trip.